### PR TITLE
feat(ora|quadbin): add Quadbin spatial indexing module for Oracle

### PR DESCRIPTION
**Pull request by diego-vicente**
# Description

Shortcut

- Story: https://app.shortcut.com/cartoteam/story/548740
- Autolink: [sc-548740]

This PR adds the Quadbin spatial indexing module to Oracle. It ports all 18 public functions (plus 2 internal helpers) from the existing BigQuery/Snowflake/Redshift implementations.

All functions are implemented as PL/SQL stored functions using bitwise arithmetic on Quadbin integer indices. Geometry functions (`QUADBIN_BOUNDARY`, `QUADBIN_CENTER`, `QUADBIN_BBOX`, `QUADBIN_POLYFILL`) use Oracle's `SDO_GEOMETRY` types. Collections are passed as JSON arrays since Oracle does not support array-typed function parameters.

**Note**: `oracle_db.py` was updated to cache the extracted wallet directory at module level, preventing TNS name invalidation when `oracledb` reuses its internal connection cache across multiple test connections (same fix as the H3 PR).

## Type of change

- Feature
- Tests

# Acceptance

73 integration tests covering all 20 functions, including edge cases for extreme latitudes, resolution boundaries (0 and 26), quadkey round-trips, sibling traversal at grid edges, polyfill with MultiPolygon/holes, and null handling.

# Basic checklist

- [X] Good PR name
- [ ] Shortcut link
- [X] Just one issue per PR
- [ ] GitHub labels
- [ ] Proper status & reviewers
- [X] Tests
- [ ] Documentation

### CODE REVIEW

File: clouds/oracle/modules/sql/quadbin/QUADBIN_POLYFILL.sql [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    v_max_tx := FLOOR(v_num_tiles * ((v_east / 360.0d) + 0.5d));
+    v_max_tx := BITAND(v_max_tx, v_num_tiles_i - 1);
+
+    -- Build result CLOB
+    DBMS_LOB.CREATETEMPORARY(v_result, TRUE);
+    DBMS_LOB.WRITEAPPEND(v_result, 1, '[');
+
+    -- Iterate all tiles in the bounding-box range
+    FOR ty IN v_min_ty .. v_max_ty LOOP
+        FOR tx IN v_min_tx .. v_max_tx LOOP
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/quadbin/QUADBIN_POLYFILL.sql]: Fixed. Switched from direct `FOR tx IN v_min_tx .. v_max_tx` to an offset-based loop: compute `v_tx_count = v_max_tx - v_min_tx + 1` from raw (unwrapped) tile coordinates, then iterate `FOR i IN 0 .. v_tx_count - 1` with `BITAND(v_min_tx + i, v_num_tiles_i - 1)` to wrap each tile index. This handles antimeridian crossings correctly because the raw coordinates preserve the spatial ordering even when the bbox spans the dateline.

File: clouds/oracle/modules/sql/quadbin/QUADBIN_FROMGEOGPOINT.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    IF point IS NULL OR resolution IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN @@ORA_SCHEMA@@.QUADBIN_FROMLONGLAT(
+        point.SDO_POINT.X,
+        point.SDO_POINT.Y,
+        resolution
+    );
+END;
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/quadbin/QUADBIN_FROMGEOGPOINT.sql:36]: Fixed. Added SDO_ORDINATES fallback -- the function now checks `point.SDO_POINT.X` first, then falls back to `point.SDO_ORDINATES(1)/SDO_ORDINATES(2)` if SDO_POINT is NULL. This matches the dual-path pattern already used in QUADBIN_POLYFILL's MBR extraction.

File: clouds/oracle/modules/sql/quadbin/QUADBIN_KRING.sql [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+RETURN VARCHAR2
+AS
+    v_zxy    VARCHAR2(200);
+    v_z      NUMBER;
+    v_x      NUMBER;
+    v_y      NUMBER;
+    v_size   NUMBER;
+    v_new_x  NUMBER;
+    v_new_y  NUMBER;
+    v_result VARCHAR2(32767);
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/quadbin/QUADBIN_KRING.sql]: Fixed. Switched `v_result` from `VARCHAR2(32767)` to `CLOB` with `DBMS_LOB` operations, matching QUADBIN_POLYFILL's pattern.

File: clouds/oracle/modules/sql/quadbin/QUADBIN_KRING_DISTANCES.sql [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+AS
+    v_zxy    VARCHAR2(200);
+    v_z      NUMBER;
+    v_x      NUMBER;
+    v_y      NUMBER;
+    v_size   NUMBER;
+    v_new_x  NUMBER;
+    v_new_y  NUMBER;
+    v_cheb   NUMBER;
+    v_result VARCHAR2(32767);
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/quadbin/QUADBIN_KRING_DISTANCES.sql]: Fixed. Same VARCHAR2 → CLOB change as QUADBIN_KRING.

File: clouds/oracle/modules/sql/quadbin/QUADBIN_TOCHILDREN.sql [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    v_current_res    NUMBER;
+    v_res_diff       NUMBER;
+    v_zoom_mask      NUMBER;
+    v_block_range    NUMBER;
+    v_sqrt_range     NUMBER;
+    v_block_shift    NUMBER;
+    v_child_base     NUMBER;
+    v_clear_mask     NUMBER;
+    v_child          NUMBER;
+    v_result         VARCHAR2(32767);
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/quadbin/QUADBIN_TOCHILDREN.sql]: Fixed. Switched to CLOB. Note: the resolution difference is already bounded by validation (`resolution < v_current_res` raises an error, and `MAX_RES = 26`), so the theoretical maximum is 4^26 children. In practice, callers use small differences (1-3 levels), but CLOB removes the hard 32KB ceiling for edge cases.

**Comment by vdelacruzb** [comment]
# Comprehensive Review — Oracle Quadbin Module

> **Updated to align with the Oracle typing convention codified in [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602).** The original typing recommendation in this comment (CLOB JSON with quoted-string BIGINTs) has been superseded. Non-typing items remain valid.

## TL;DR

Solid work overall. Algorithm parity with BQ/SF/DB is strong, happy-path test values match element-by-element, and SRID 4326 is set correctly on geometry outputs.

To land cleanly, the quadbin module needs to be refactored to the agreed Oracle typing convention (nested table types + pipelined functions for arrays/struct-arrays, object types for structs). The existing docs, SRID, and test-infrastructure gaps are independent of typing and apply unchanged.

---

## 1. Type Strategy — align with [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602)

The agreed convention for Oracle SQL-native returns:

| BQ/SF shape | Oracle type |
|---|---|
| `ARRAY<BIGINT>` | `TABLE OF NUMBER` PIPELINED |
| `ARRAY<STRING>` | `TABLE OF VARCHAR2(...)` PIPELINED |
| `ARRAY<STRUCT<...>>` | `TABLE OF <object_type>` PIPELINED |
| `STRUCT<...>` | `OBJECT` type with named fields |
| scalar `GEOMETRY` | `SDO_GEOMETRY` with explicit SRID 4326 |

Consumer idiom is native Oracle SQL: `SELECT column_value FROM TABLE(fn(...))` or `SELECT t.field FROM TABLE(fn(...)) t`. No `JSON_TABLE`, no CARTO-provided wrappers.

### Quadbin-specific refactor (P0)

| File | Current | Target |
|---|---|---|
| `clouds/oracle/modules/sql/quadbin/00_types.sql` | — | **Create** with `QUADBIN_INDEX_ARRAY` (TABLE OF NUMBER), `QUADBIN_ZXY` (OBJECT z/x/y), `QUADBIN_BBOX` (OBJECT west/south/east/north), `QUADBIN_DISTANCE_PAIR` (OBJECT index/distance), `QUADBIN_DISTANCE_ARRAY` (TABLE OF QUADBIN_DISTANCE_PAIR) |
| `QUADBIN_KRING.sql` | CLOB JSON | `QUADBIN_INDEX_ARRAY` PIPELINED |
| `QUADBIN_TOCHILDREN.sql` | CLOB JSON | `QUADBIN_INDEX_ARRAY` PIPELINED |
| `QUADBIN_POLYFILL.sql` | CLOB JSON | `QUADBIN_INDEX_ARRAY` PIPELINED |
| `QUADBIN_KRING_DISTANCES.sql` | CLOB JSON of objects | `QUADBIN_DISTANCE_ARRAY` PIPELINED |
| `QUADBIN_TOZXY.sql` | VARCHAR2 JSON | `QUADBIN_ZXY` object (non-pipelined scalar return) |
| `QUADBIN_BBOX.sql` | VARCHAR2 JSON array | `QUADBIN_BBOX` object (non-pipelined scalar return) |
| All `test_QUADBIN_*.py` for affected functions | `json.loads(cur.fetchone()[0])` | `SELECT ... FROM TABLE(fn(...))` or `.field` access |

### Items no longer applicable (dropped from TODO)

- ~~Quote BIGINTs as strings in JSON output~~ — under `TABLE OF NUMBER`, BIGINTs traverse the Oracle driver natively at full 38-digit precision. No JSON, no JavaScript precision cliff.
- ~~Regression test asserting quoted-JSON content~~ — not applicable under nested tables.

---

## 2. Documentation — still missing (P1)

[clouds/oracle/modules/doc/quadbin/](clouds/oracle/modules/doc/quadbin/) does not exist.

- BigQuery: 22 `.md` files + `_INTRO.md` + `images/`
- Snowflake: 19 `.md` files + `_INTRO.md`
- Oracle: nothing

Need all 18 function `.md` docs + `_INTRO.md` mirroring the Snowflake structure, adapted for Oracle types and the `TABLE(...)` consumer idiom.

---

## 3. SRID Handling — correct outputs, commentary gap on inputs

✓ Geometry-returning functions set SRID 4326 explicitly:
- [`QUADBIN_BOUNDARY.sql:34`](clouds/oracle/modules/sql/quadbin/QUADBIN_BOUNDARY.sql#L34): `SDO_GEOMETRY(2003, 4326, ...)`
- [`QUADBIN_CENTER.sql:46`](clouds/oracle/modules/sql/quadbin/QUADBIN_CENTER.sql#L46): `SDO_GEOMETRY(2001, 4326, ...)`

**Input-side (P2)**: [`QUADBIN_FROMGEOGPOINT.sql`](clouds/oracle/modules/sql/quadbin/QUADBIN_FROMGEOGPOINT.sql) doesn't validate input SRID. Either add a comment documenting the WGS84 assumption or raise if `g.SDO_SRID != 4326`. Same for [`QUADBIN_FROMLONGLAT.sql`](clouds/oracle/modules/sql/quadbin/QUADBIN_FROMLONGLAT.sql).

---

## 4. Test Parity — happy paths match; error semantics align with Snowflake (not a divergence)

### Happy paths ✓

All spot-checked expected values match BQ/SF element-by-element. Examples: `FROMZXY(4, 9, 8)` → 5209574053332910079; `TOPARENT(5209574053332910079, 3)` → 5205105638077628415. KRING, KRING_DISTANCES, BOUNDARY, CENTER all match.

### Error semantics — Oracle matches Snowflake, not a bug

Initial review flagged this as a P0 divergence; that was incorrect. Snowflake also returns NULL on invalid inputs:

| Input | BigQuery | Snowflake | Oracle |
|---|---|---|---|
| `TOPARENT(idx, -1)` | throws | NULL (via `IFF` in SQL) | NULL |
| `TOPARENT(idx, 27)` | throws | NULL | NULL |
| `TOPARENT(idx, res > idx_res)` | throws | returns garbage (no check) | NULL ← Oracle stricter |

Oracle's NULL-on-invalid matches Snowflake's native-SQL convention (the correct peer — BigQuery throws only because its UDFs are JS-based). Oracle is actually *stricter* than Snowflake on the "resolution > index's own resolution" case — a good divergence. This aligns with the [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602) NULL-on-invalid rule.

### Coverage

Oracle: 82 test cases across 20 functions. BigQuery: 54. Snowflake: 38. Oracle has more coverage than either peer.

### Test infrastructure (P1)

All 20 Oracle quadbin tests use the legacy import pattern:

```python
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'common'))
from run_query import run_query
```

Should migrate to the canonical pattern (per [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602) and the existing LDS module):

```python
from test_utils import run_query, drop_table
```

Removes the `sys.path` hack and exposes `drop_table` (useful for future table-fixture tests).

---

## Prioritized Issue List

### P0 — blocking
1. **Refactor array/struct returns to nested tables per [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602)** — create `00_types.sql`, refactor 6 functions (`KRING`, `TOCHILDREN`, `POLYFILL`, `KRING_DISTANCES`, `TOZXY`, `BBOX`), migrate tests to `TABLE(...)` / field-access pattern
2. Add `oracle` to `VALID_CLOUDS` in `core/Makefile` (shared blocker with [#594](https://github.com/CartoDB/analytics-toolbox-core/pull/594))

### P1 — strongly recommended before merge
3. Create [`clouds/oracle/modules/doc/quadbin/`](clouds/oracle/modules/doc/quadbin/) with 18 function `.md` files + `_INTRO.md`
4. Migrate all 20 test files to `from test_utils import run_query, drop_table` (drop `sys.path.insert`)
5. Rebase on [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602) once merged, or fold the rules update into this PR

### P2 — follow-up acceptable
6. Add SRID comments/validation to [`QUADBIN_FROMGEOGPOINT.sql`](clouds/oracle/modules/sql/quadbin/QUADBIN_FROMGEOGPOINT.sql) and [`QUADBIN_FROMLONGLAT.sql`](clouds/oracle/modules/sql/quadbin/QUADBIN_FROMLONGLAT.sql)
7. Change `QUADBIN_ISVALID` return type from bare `NUMBER` to `NUMBER(1)` per the [#602](https://github.com/CartoDB/analytics-toolbox-core/pull/602) convention (single-digit constraint at the schema level); audit other quadbin functions for boolean-returning helpers using bare `NUMBER` and apply the same change
8. Document the `NUMBER(1)` (1/0) boolean convention for `QUADBIN_ISVALID` — callers use `= 1`
9. Add POLYFILL edge-case tests: degenerate geometries, antimeridian-crossing polygons, large polygons
10. Consider extracting `_GENERATE_RANGE` / `_QUADBIN_KRING` helpers (SF-style)
11. Document `QUADBIN_POLYFILL_TABLE` omission (BQ-only) in module notes

### Verified non-issues
- ✓ SRID 4326 set explicitly on BOUNDARY and CENTER outputs
- ✓ Happy-path expected test values match BQ/SF element-by-element
- ✓ Error semantics (NULL-on-invalid) match Snowflake — a feature of the convention, not a divergence from BQ
- ✓ Function coverage matches BQ/SF (18 public + 2 internal; QUADBIN_POLYFILL_TABLE intentionally BQ-only)
- ✓ `NUMBER` datatype adequately represents 64-bit integers and floats at compute time

**Comment by vdelacruzb** [comment]
### Pending to fix

Administrative items to resolve before merge:

- [ ] **Shortcut ticket** — PR description still has placeholder `STORY_ID` text in the Shortcut section; a real story needs to be linked
- [ ] **Epic assignment** — no epic associated

These are independent of the code content of the PR.

**Comment by diego-vicente** [comment]
Thanks for the comprehensive review. All P0/P1 items and most P2 are addressed in the latest commits (not yet pushed). Summary by priority:

### P0 — done
1. **Nested-table refactor**
   - New `clouds/oracle/modules/sql/quadbin/00_types.sql` defines `QUADBIN_ZXY`, `QUADBIN_BBOX_OBJ`, `QUADBIN_INDEX_ARRAY`, `QUADBIN_DISTANCE_PAIR`, `QUADBIN_DISTANCE_ARRAY` (with idempotent `DROP … FORCE` blocks).
   - **Naming deviation:** I used `QUADBIN_BBOX_OBJ` rather than `QUADBIN_BBOX` because the function `QUADBIN_BBOX` already occupies that name in Oracle's shared type/function namespace. Happy to revisit if you have a different preference.
   - `KRING`, `TOCHILDREN`, `POLYFILL` → PIPELINED `QUADBIN_INDEX_ARRAY`; `KRING_DISTANCES` → PIPELINED `QUADBIN_DISTANCE_ARRAY`; `TOZXY` → `QUADBIN_ZXY`; `BBOX` → `QUADBIN_BBOX_OBJ`.
   - Downstream callers (`BOUNDARY`, `CENTER`, `DISTANCE`, `SIBLING`) updated to consume objects via field access; no more `JSON_VALUE`.
   - `TOCHILDREN` and `POLYFILL` aligned to NULL-on-invalid (empty pipeline) per the convention.
   - Affected tests rewritten to `SELECT … FROM TABLE(fn(...))` / `t.field` access.
2. `oracle` added to `VALID_CLOUDS` in root `Makefile`.

### P1 — done
3. `clouds/oracle/modules/doc/quadbin/` created with 18 function `.md` files + `_INTRO.md`, mirroring the Snowflake structure with Oracle types and the `TABLE(...)` consumer idiom.
4. All 20 tests migrated to `from test_utils import run_query` (and `drop_table` available); `sys.path.insert` and `from run_query import run_query` are gone, `fetch=True` removed (the `test_utils` helper auto-fetches).
5. Rebased onto current `main` (which now contains #602).

### P2 — done
6. `QUADBIN_FROMGEOGPOINT` and `QUADBIN_FROMLONGLAT` now document the WGS84 (EPSG:4326) input assumption explicitly. No auto-transform in v1.0 per the convention.
7. `QUADBIN_ISVALID` documented as conventional `NUMBER(1)` 1/0. PL/SQL function return types cannot carry the precision constraint (Oracle requires unconstrained types on `RETURN`), so the contract lives in the doc/comment; the function body still returns 0/1 only.
8. The `NUMBER(1)` 1/0 boolean convention and `= 1` callsite idiom are described in `QUADBIN_ISVALID.md`.
9. POLYFILL edge-case tests added: degenerate point, antimeridian-crossing polygon, large bbox, invalid resolution.
11. `QUADBIN_POLYFILL_TABLE` omission noted in `_INTRO.md` (Oracle's pipelined `QUADBIN_POLYFILL` already supports `INSERT … SELECT` materialization, so the BQ-only sproc has no equivalent).

### P2 — not done
10. Helper extraction (`_GENERATE_RANGE` / `_QUADBIN_KRING`): skipped to keep the diff scoped. Happy to follow up in a separate PR if you want it.

### Bonus
- `clouds/oracle/common/build_modules.js` now `.sort()`s `readdirSync` results so the `00_` prefix actually guarantees deploy ordering on Linux CI (it had been relying on filesystem-dependent natural order).

**Review by vdelacruzb — APPROVED** [review]
I think we are good to go on this one. I've refactored the QUADBIN_POLYFILL to match the implementation in SF. And was working fine in one case that was giving troubles last month in Databricks.
Image

Tried approaches like the one in BQ but it wasn't working properly.
